ecmascript/minifier/src/pass/mangle_props.rs

```diff
@@ -1,10 +1,211 @@
+use crate::option::ManglePropertiesOptions;
+use crate::util::base54::base54;
+use std::collections::{HashMap, HashSet};
+use std::rc::Rc;
+use swc_ecma_ast::{CallExpr, Expr, ExprOrSuper, Ident, Lit, MemberExpr, Module, PropName, Str};
+use swc_ecma_visit::{VisitMut, VisitMutWith};
 
-pub fn property_mangler() -> impl VisitMut {
-    Mangler {}
+#[derive(Debug, Default)]
+struct ManglePropertiesState {
+    options: ManglePropertiesOptions,
+
+    names_to_mangle: HashSet<String>,
+    unmangleable: HashSet<String>,
+
+    // Cache of already mangled names
+    cache: HashMap<String, String>,
+}
+
+impl ManglePropertiesState {
+    fn add(&mut self, name: String) {
+        if self.can_mangle(&name) {
+            self.names_to_mangle.insert(name.clone());
+        }
+
+        if !self.should_mangle(&name) {
+            self.unmangleable.insert(name.clone());
+        }
+    }
+
+    fn can_mangle(&self, name: &str) -> bool {
+        if self.unmangleable.contains(name) {
+            false
+        } else if self.options.reserved.contains(&name.to_string()) {
+            false
+        } else {
+            // TODO only_cache, check if it's a name that doesn't need quotes
+            true
+        }
+    }
+
+    fn should_mangle(&self, name: &str) -> bool {
+        if self.options.reserved.contains(&name.to_string()) {
+            false
+        } else {
+            self.cache.contains_key(name) || self.names_to_mangle.contains(name)
+        }
+        // TODO check regex option,
+    }
+}
+
+pub fn mangle_properties<'a>(m: &mut Module, options: ManglePropertiesOptions) {
+    let mut state = ManglePropertiesState {
+        options,
+        ..Default::default()
+    };
+
+    m.visit_mut_with(&mut PropertyCollector { state: &mut state });
+
+    m.visit_mut_with(&mut Mangler {
+        state: &mut state,
+        n: 0,
+    });
+}
+
+#[derive(Debug)]
+struct Mangler<'a> {
+    state: &'a mut ManglePropertiesState,
+    n: usize,
+}
+
+impl Mangler<'_> {
+    fn mangle(&mut self, name: &str) -> Option<String> {
+        if self.state.should_mangle(name) {
+            if let Some(cached) = self.state.cache.get(name) {
+                Some(cached.clone())
+            } else {
+                let n = self.n;
+                self.n += 1;
+                let mangled_name = base54(n);
+                self.state
+                    .cache
+                    .insert(String::from(name), mangled_name.clone());
+                Some(mangled_name)
+            }
+        } else {
+            None
+        }
+    }
+
+    fn mangle_ident(&mut self, ident: &mut Ident) {
+        if let Some(mangled) = self.mangle(ident.sym.as_ref()) {
+            ident.sym = mangled.into();
+        }
+    }
+
+    fn mangle_str(&mut self, string: &mut Str) {
+        if let Some(mangled) = self.mangle(&string.value) {
+            string.value = mangled.into();
+        }
+    }
+
+    // TODO mangle_private
+}
+
+impl VisitMut for Mangler<'_> {
```

Comment:
I think `Prop::Shorthand` should be handled.
You can do this by cloning the original identifier first and using it as a key of `Prop::KeyValue`.